#1. 개요
> * dataset은 370개의 feature로 이루어진 데이터
* 산탄데르 은행에서 캐글에 경연 의뢰한 데이터로 피처 이름은 임의로 처리됨
* TARGET 정보가 1이면 불만족, 0이면 만족
* https://www.kaggle.com/c/santander-customer-satisfaction/data

#2. 파일 읽기

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd

train_file = '/content/drive/MyDrive/3. 머신러닝/6. 앙상블/data/santander-customer-satisfaction/train.csv'
test_file = '/content/drive/MyDrive/3. 머신러닝/6. 앙상블/data/santander-customer-satisfaction/test.csv'
df_train = pd.read_csv(train_file)
df_test = pd.read_csv(test_file)

df_train.head(2)

ID  var3  var15  ...  saldo_medio_var44_ult3     var38  TARGET
0   1     2     23  ...                     0.0  39205.17       0
1   3     2     34  ...                     0.0  49278.03       0

[2 rows x 371 columns]

#3. 데이터 전처리

##3.1 데이터 확인

In [ ]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76020 entries, 0 to 76019
Columns: 371 entries, ID to TARGET
dtypes: float64(111), int64(260)
memory usage: 215.2 MB


In [ ]:
df_train.describe()

ID           var3  ...         var38        TARGET
count   76020.000000   76020.000000  ...  7.602000e+04  76020.000000
mean    75964.050723   -1523.199277  ...  1.172358e+05      0.039569
std     43781.947379   39033.462364  ...  1.826646e+05      0.194945
min         1.000000 -999999.000000  ...  5.163750e+03      0.000000
25%     38104.750000       2.000000  ...  6.787061e+04      0.000000
50%     76043.000000       2.000000  ...  1.064092e+05      0.000000
75%    113748.750000       2.000000  ...  1.187563e+05      0.000000
max    151838.000000     238.000000  ...  2.203474e+07      1.000000

[8 rows x 371 columns]

In [ ]:
df_test.describe()

ID           var3  ...  saldo_medio_var44_ult3         var38
count   75818.000000   75818.000000  ...            75818.000000  7.581800e+04
mean    75874.830581   -1579.955011  ...               70.504319  1.173863e+05
std     43882.370827   39752.473358  ...             3318.527783  2.479384e+05
min         2.000000 -999999.000000  ...                0.000000  1.202730e+03
25%     37840.250000       2.000000  ...                0.000000  6.754961e+04
50%     75810.000000       2.000000  ...                0.000000  1.057643e+05
75%    113996.500000       2.000000  ...                0.000000  1.183159e+05
max    151837.000000     238.000000  ...           354260.730000  2.889440e+07

[8 rows x 370 columns]

>var3의 최소값이 -999999로 무언가 정상적이지 않다고 생각할 수 있다.

##3.2 이상치 수정

In [ ]:
df_train['var3'].value_counts().sort_index()

-999999      116
 0            75
 1           105
 2         74165
 3           108
           ...  
 228           1
 229           1
 231           1
 235           1
 238           1
Name: var3, Length: 208, dtype: int64

In [ ]:
df_test['var3'].value_counts().sort_index()

-999999      120
 0            76
 1            99
 2         73962
 3           107
           ...  
 227           2
 229           1
 230           2
 233           1
 238           1
Name: var3, Length: 199, dtype: int64

> var3를 기준으로 훈련 데이터와 검증데이터에 각각 120개 가량의 데이터가 존재한다. 그리고 바로 다음 수치는 0으로 나타난다. 이는 특정 에러가 발생했을 경우 나타났을 것으로 판단하며(보통 개발 시 절대 나타날 수 없는 값을 에러로 사용하는 경우도 있음으로) 가장 빈도가 많은 2로 대체할 예정이다.(4분위수 대다수에 포진됨)

> -999999가 116개 나타난 것을 볼 수 있다. 여기서는 최빈값으로 값을 대체하자

In [ ]:
df_train['var3'].replace(-999999,2, inplace=True)
df_test['var3'].replace(-999999,2, inplace=True)
df_train['var3'].value_counts().sort_index()

0         75
1        105
2      74281
3        108
4         86
       ...  
228        1
229        1
231        1
235        1
238        1
Name: var3, Length: 207, dtype: int64

##3.3 불필요한 feature 제거

In [ ]:
df_train.describe()

ID          var3  ...         var38        TARGET
count   76020.000000  76020.000000  ...  7.602000e+04  76020.000000
mean    75964.050723      2.716483  ...  1.172358e+05      0.039569
std     43781.947379      9.447971  ...  1.826646e+05      0.194945
min         1.000000      0.000000  ...  5.163750e+03      0.000000
25%     38104.750000      2.000000  ...  6.787061e+04      0.000000
50%     76043.000000      2.000000  ...  1.064092e+05      0.000000
75%    113748.750000      2.000000  ...  1.187563e+05      0.000000
max    151838.000000    238.000000  ...  2.203474e+07      1.000000

[8 rows x 371 columns]

> 첫번째 feature인 ID는 단순 식별자 일 것임으로 예측에 불필요한 데이터가 된다. 

In [ ]:
df_train.drop('ID',axis=1 , inplace=True)
df_test.drop('ID',axis=1 , inplace=True)
df_train.head()

var3  var15  ...          var38  TARGET
0     2     23  ...   39205.170000       0
1     2     34  ...   49278.030000       0
2     2     23  ...   67333.770000       0
3     2     37  ...   64007.970000       0
4     2     39  ...  117310.979016       0

[5 rows x 370 columns]

#4. 데이터 분리

In [ ]:
y_train = df_train['TARGET']
X_train = df_train.drop(columns=['TARGET'], axis=1)

y_test = df_train['TARGET']
X_test = df_train.drop(columns=['TARGET'], axis=1)

#5. XGBoost

In [ ]:
from xgboost import XGBClassifier
from sklearn.metrics import roc_auc_score

# n_estimators는 500으로, random state는 예제 수행 시마다 동일 예측 결과를 위해 설정. 
xgb_clf = XGBClassifier(n_estimators=500, random_state=62)

# 성능 평가 지표를 auc로, 조기 중단 파라미터는 100으로 설정하고 학습 수행. 
xgb_clf.fit(X_train, y_train, early_stopping_rounds=100,
            eval_metric="auc", eval_set=[(X_train, y_train), (X_test, y_test)])

- https://developers.google.com/machine-learning/crash-course/classification/roc-and-auc?hl=ko
- https://bskyvision.com/1165

In [ ]:
xgb_clf.predict_proba(X_test)

array([[0.9664806 , 0.03351941],
       [0.9892949 , 0.0107051 ],
       [0.9975809 , 0.00241912],
       ...,
       [0.99550766, 0.00449237],
       [0.98925155, 0.01074846],
       [0.81238544, 0.18761459]], dtype=float32)

In [ ]:
y_test

0        0
1        0
2        0
3        0
4        0
        ..
76015    0
76016    0
76017    0
76018    0
76019    0
Name: TARGET, Length: 76020, dtype: int64

In [ ]:
xgb_roc_score = roc_auc_score(y_test, xgb_clf.predict_proba(X_test)[:,1])
print('ROC AUC: {0:.4f}'.format(xgb_roc_score))

ROC AUC: 0.8835


> AUC는 이진데이터 구분 시 사용되며 만족 불만족 두 가지에 대한 평가를 하여 사용하였다.

> predict_proba는 확률을 얻게 되며 보통 1이 될 확률로 값을 구하게 된다.

In [ ]:
from sklearn.model_selection import GridSearchCV

# 하이퍼 파라미터 테스트의 수행 속도를 향상시키기 위해 n_estimators를 100으로 감소
xgb_clf = XGBClassifier(n_estimators=100)

# min_child_weight는 min_samples_leaf 비슷한 값으로 과적합 해결
# colsample_bytree는 max_features와 유사하며 feature가 많을 경우 과적합 조절을 위해 사용, 확률 데이터
params = {'max_depth':[3, 6] , 'min_child_weight':[1,3] ,'colsample_bytree':[0.5, 0.75] }

# cv는 3으로 지정 
gridcv = GridSearchCV(xgb_clf, param_grid=params, cv=3)
gridcv.fit(X_train, y_train, early_stopping_rounds=30, eval_metric="auc",
           eval_set=[(X_test, y_test)])

print('GridSearchCV 최적 파라미터:',gridcv.best_params_) 

xgb_roc_score = roc_auc_score(y_test, gridcv.predict_proba(X_test)[:,1])
print('ROC AUC: {0:.4f}'.format(xgb_roc_score))

[0]	validation_0-auc:0.704403
Will train until validation_0-auc hasn't improved in 30 rounds.
[1]	validation_0-auc:0.797704
[2]	validation_0-auc:0.79624
[3]	validation_0-auc:0.801781
[4]	validation_0-auc:0.808353
[5]	validation_0-auc:0.805543
[6]	validation_0-auc:0.80468
[7]	validation_0-auc:0.808946
[8]	validation_0-auc:0.813852
[9]	validation_0-auc:0.813157
[10]	validation_0-auc:0.814964
[11]	validation_0-auc:0.814724
[12]	validation_0-auc:0.82148
[13]	validation_0-auc:0.816273
[14]	validation_0-auc:0.816444
[15]	validation_0-auc:0.819221
[16]	validation_0-auc:0.820738
[17]	validation_0-auc:0.823917
[18]	validation_0-auc:0.826145
[19]	validation_0-auc:0.824276
[20]	validation_0-auc:0.825362
[21]	validation_0-auc:0.82302
[22]	validation_0-auc:0.821821
[23]	validation_0-auc:0.825328
[24]	validation_0-auc:0.826756
[25]	validation_0-auc:0.825068
[26]	validation_0-auc:0.823808
[27]	validation_0-auc:0.82599
[28]	validation_0-auc:0.828181
[29]	validation_0-auc:0.827815
[30]	validation_0-auc

15분 소요

In [ ]:
import numpy as np

# n_estimators는 1000으로 증가시키고, learning_rate=0.02로 감소
# reg_alpha : 피처 개수가 많을 때 적용을 검토, 클수록 과적합 감소 효과
xgb_clf = XGBClassifier(n_estimators=1000, random_state=156, learning_rate=0.02, max_depth=6, min_child_weight=1, colsample_bytree=0.5, reg_alpha=0.03)

# 조기 중단 파라미터 값은 200으로 설정
xgb_clf.fit(X_train, y_train, early_stopping_rounds=200, eval_metric="auc", eval_set=[(X_train, y_train), (X_test, y_test)])

xgb_roc_score = roc_auc_score(y_test, xgb_clf.predict_proba(X_test)[:, 1])
print(f'ROC-AUC: {np.round(xgb_roc_score, 4)}')

[0]	validation_0-auc:0.820417	validation_1-auc:0.820417
Multiple eval metrics have been passed: 'validation_1-auc' will be used for early stopping.

Will train until validation_1-auc hasn't improved in 200 rounds.
[1]	validation_0-auc:0.823881	validation_1-auc:0.823881
[2]	validation_0-auc:0.824678	validation_1-auc:0.824678
[3]	validation_0-auc:0.831163	validation_1-auc:0.831163
[4]	validation_0-auc:0.834837	validation_1-auc:0.834837
[5]	validation_0-auc:0.830563	validation_1-auc:0.830563
[6]	validation_0-auc:0.834569	validation_1-auc:0.834569
[7]	validation_0-auc:0.836347	validation_1-auc:0.836347
[8]	validation_0-auc:0.837404	validation_1-auc:0.837404
[9]	validation_0-auc:0.839547	validation_1-auc:0.839547
[10]	validation_0-auc:0.83836	validation_1-auc:0.83836
[11]	validation_0-auc:0.837228	validation_1-auc:0.837228
[12]	validation_0-auc:0.836121	validation_1-auc:0.836121
[13]	validation_0-auc:0.837369	validation_1-auc:0.837369
[14]	validation_0-auc:0.839224	validation_1-auc:0.839224

KeyboardInterrupt: ignored

6분 정도 소요

>LightGBM

In [ ]:
from lightgbm import LGBMClassifier

lgbm_clf = LGBMClassifier(n_estimators=500)

evals = [(X_test, y_test)]
lgbm_clf.fit(X_train, y_train, early_stopping_rounds=100, eval_metric='auc', eval_set=evals, verbose=True)

lgbm_roc_score = roc_auc_score(y_test, lgbm_clf.predict_proba(X_test)[:, 1], average='macro')
print(f'ROC-AUC: {np.round(lgbm_roc_score, 4)}')

튜닝

In [ ]:
from sklearn.model_selection import GridSearchCV

# 하이퍼 파라미터 테스트의 수행 속도를 향상시키기위해 n_estimators=200 으로 설정
lgbm_clf = LGBMClassifier(n_estimators=200)

params = {'num_leaves': [32, 64],
         'max_depth': [128, 160],
         'min_child_samples': [60, 100],
         'subsample': [0.8, 1]}

gridcv = GridSearchCV(lgbm_clf, param_grid=params, cv=3)
gridcv.fit(X_train, y_train, early_stopping_rounds=30, eval_metric='auc', eval_set=[(X_train, y_train), (X_test, y_test)])

print(f'GridSearchCV 최적 파라미터: {gridcv.best_params_}')
lgbm_roc_score = roc_auc_score(y_test, gridcv.predict_proba(X_test)[:, 1], average='macro')
print(f'ROC-AUC: {np.round(lgbm_roc_score, 4)}')

4분

In [ ]:
lgbm_clf = LGBMClassifier(n_estimators=1000,max_depth=128, min_child_samples=100, num_leaves=32, subsample=0.8)

evals = [(X_test, y_test)]
lgbm_clf.fit(X_train, y_train, early_stopping_rounds=100, eval_metric='auc', eval_set=evals, verbose=True)

lgbm_roc_score = roc_auc_score(y_test, lgbm_clf.predict_proba(X_test)[:, 1], average='macro')
print(f'ROC-AUC: {np.round(lgbm_roc_score, 4)}')